### PR TITLE
Deploy App Service Plan - MLZ Example

### PR DESCRIPTION
**Pull request by ExchMaster**
# Description

This update adds a new example workload deployment.  Specifically, it enables the deployment of an App Service Plan suitable for hosting Linux web applications using docker containers.   Here are some of the details of the implementation:

- Updates mlz.bicep to include resourcePrefix as an deployment output.  This allows examples to pick up the resourcePrefix straight from the deploymentVariables file.   User can specify their own name via param if desired.
- Modified gitignore by adding a single line exception permitting files with the name 'plan' and in the ./src/bicep/* to be committed via git.
- Enabled the ability for the user doing the deployment to use custom resource group and specific subscription if desired.  If unspecified, the default behavior is to deploy into an appropriate MLZ resource group with corresponding subscription.  In this case, the shared services resource group.

Note:  This example does not also deploy a web application because the target audience for this particular sample is the engineering/admin team.   A separate example showing how to deploy a Linux web app container on docker will be forthcoming which is aimed at the dev team.

This update uses the bicep shared variable file pattern as described here:  https://docs.microsoft.com/en-us/azure/azure-resource-manager/bicep/patterns-shared-variable-file

## Issue reference

This PR is related to the following issue: #517
I did attempt to retrieve MLZ outputs using a deployment object, but ultimately failed to successfully do so do the dynamic nature of how the outputs are parsed/discovered at run-time.   More research is needed, but felt this pattern solved the majority of the issue and is worthy of doing now while we continue to investigate how to use deployment objects for this problem.

# Checklist

Please make sure you've completed the relevant tasks for this PR out of the following list:

* [see above] All acceptance criteria in the backlog item are met
* [ *] The documentation is updated to cover any new or changed features
* [ *] Manual tests have passed
* [ *] Relevant issues are linked to this PR

### CODE REVIEW

File: src/bicep/examples/appServicePlan/appService.bicep [hunk 1]
```diff
@@ -0,0 +1,51 @@
...
+param enableAutoScale bool = true
+
+var mlzDeploymentVariables = json(loadTextContent('../deploymentVariables.json'))
+var appServicePlanName_Var = contains(appServicePlanName, 'placeHolder') ? '${mlzDeploymentVariables.mlzResourcePrefix.Value}--ASP' : appServicePlanName
+var targetResourceGroup_Var = contains(targetResourceGroup, 'placeHolder') ? '${mlzDeploymentVariables.spokes.Value[2].resourceGroupName}' : targetResourceGroup
+var targetSubscriptionId_Var = contains(targetResourceGroup, 'placeHolder') ? '${mlzDeploymentVariables.spokes.Value[2].subscriptionId}' : subscription().subscriptionId
+var location = deployment().location
+var kind = 'linux'
+var capacity = 2
+var sku = 'premium'
```
Comment (glennmusa) [inline src/bicep/examples/appServicePlan/appService.bicep:23]: We can simplify these ternary var assignments on L16-L20 by removing them and allowing for the default parameter override behavior of ARM and Bicep to perform that same evaluation of "Did they specify a parameter value? Okay use that. Did they not specify that parameter? Okay, use the mlzDeploymentVaraibles property instead."

My suggestion below may not be complete, but this is the general idea:

```suggestion
@description('A deploymentVariables.json file that contains MLZ deployment outputs.')
param mlzDeploymentVariables object = json(loadTextContent('../deploymentVariables.json'))

@description('The name of the web server farm which will be created. If unchanged or not specified, the MLZ resource prefix + "-ASP" will be utilized.')
param appServicePlanName string = '${mlzDeploymentVariables.mlzResourcePrefix.Value}-ASP'

@description('The name of the resource group in which the app service plan will be deployed. If unchanged or not specified, the MLZ shared services resource group is used.')
param targetResourceGroupName string = '${mlzDeploymentVariables.spokes.Value[2].resourceGroupName}'

@description('The subscription ID of the resource group in which the app service plan will be deployed. If unchanged or not specified, the MLZ shared services subscription is used.')
param targetSubscriptionId string = '${mlzDeploymentVariables.spokes.Value[2].subscriptionId}'

@description('If true, enables dynamic scale-in & scale-out based on CPU percentages. If false, then compute instances remain static with 2 instances supporting all traffic')
param enableAutoScale bool = true

@description('The region to deploy the app service plan resources into. If unchanged or not specified, the deployment location will be used.')
param location string = deployment().location

var kind = 'linux'
var capacity = 2
var sku = 'premium'
```

File: src/bicep/examples/appServicePlan/modules/appServiceSettings.bicep [hunk 2]
```diff
@@ -0,0 +1,92 @@
...
+  name: settingName_var
+  location: location
+  properties: {
+    profiles: [
+      {
+        name: 'DefaultAutoscaleProfile'
+        capacity: {
+          minimum: minimumCapacity
+          maximum: maximumCapacity
+          default: defaultCapacity
```
Comment (glennmusa) [inline src/bicep/examples/appServicePlan/modules/appServiceSettings.bicep:47]: I agree with accepting typed inputs, but some of the properties in this API expect strings:

```suggestion
          minimum: string(minimumCapacity)
          maximum: string(maximumCapacity)
          default: string(defaultCapacity)
```

File: src/bicep/examples/appServicePlan/modules/appServiceSettings.bicep [hunk 3]
```diff
@@ -0,0 +1,92 @@
...
+              statistic: 'Average'
+              timeWindow: 'PT10M'
+              timeAggregation: 'Average'
+              operator: 'GreaterThan'
+              threshold: metricThresholdToScaleOut
+            }
+            scaleAction: {
+              direction: 'Increase'
+              type: 'PercentChangeCount'
+              value: changePercentScaleOut
```
Comment (glennmusa) [inline src/bicep/examples/appServicePlan/modules/appServiceSettings.bicep:64]:
```suggestion
              value: string(changePercentScaleOut)
```

File: src/bicep/examples/appServicePlan/modules/appServiceSettings.bicep [hunk 4]
```diff
@@ -0,0 +1,92 @@
...
+              statistic: 'Average'
+              timeWindow: 'PT10M'
+              timeAggregation: 'Average'
+              operator: 'LessThan'
+              threshold: metricThresholdToScaleIn
+            }
+            scaleAction: {
+              direction: 'Decrease'
+              type: 'PercentChangeCount'
+              value: changePercentScaleIn
```
Comment (glennmusa) [inline src/bicep/examples/appServicePlan/modules/appServiceSettings.bicep:82]:
```suggestion
              value: string(changePercentScaleIn)
```

File: src/bicep/examples/appServicePlan/README.md [hunk 5]
```diff
@@ -0,0 +1,59 @@
...
+
+One way to generate the MLZ variable file(deploymentVariables.json) which contains all of the needed values for this examples and others as well is through PowerShell Core and the Azure PowerShell module.  Both PowerShell Core and the Azure PowerShell module are open source projects and avaliable for all major operating systems (Mac, Linux, Windows).
+
+* Get PowerShell Core:  <https://github.com/PowerShell/PowerShell/releases>
+* Get Azure PowerShell: <https://docs.microsoft.com/en-us/powershell/azure/install-az-ps>
+* Getting Started with Azure PowerShell: <https://docs.microsoft.com/en-us/powershell/azure/get-started-azureps>
+* Generate 'deploymentVariables.json': (Get-AzSubscriptionDeployment -Name MLZDeploymentName).outputs | ConvertTo-Json -Depth 10 | Out-File -FilePath .\deploymentVariables.json
+* Replace "MLZDeploymentName" with your deployment name:  Browse to 'Subscriptions', to the subscription MLZ was deployed into, and then look at 'Deployments'
+
+Place the 'deploymentVariables.json' file ./src/bicep/examples folder.  See the sample for reference.
```
Comment (glennmusa) [inline src/bicep/examples/appServicePlan/README.md:40]: Let's promote the instructions to generate a deployment variables file to the `src/bicep/examples` README and link out to it from this sample. We should keep the `see the sample deployment variables file` bit and update the steps in `Deploying App Service Plan`.

File: .gitignore [hunk 6]
```diff
@@ -16,6 +16,7 @@ crash.log
 # Include tfplan files to ignore the plan output of command: terraform plan -out=tfplan
 # example: *tfplan*
 *plan*
+!src/bicep/**
```
Comment (glennmusa) [inline .gitignore:19]: Is this change intentional?

File: src/bicep/examples/appServicePlan/modules/appServiceSettings.bicep [hunk 7]
```diff
@@ -0,0 +1,92 @@
+
+
```
Comment (glennmusa) [inline src/bicep/examples/appServicePlan/modules/appServiceSettings.bicep:2]: nit: whitespace
```suggestion
```

**Review by glennmusa — CHANGES_REQUESTED** [review]
Looking good! Some minor nits and questions. My primary feedback is for allowing for the default parameter override behavior of ARM and Bicep. 👍

File: .gitignore (hunk 6 above)
Comment (ExchMaster) [reply .gitignore:19]: Yes sir.   The !src/bicep/** creates an exception to the preceding line before it, "*plan*".  The exception is required because the name of the Azure service I am trying to implement is "App Service Plan".  Git was ignoring all the files in this commit because of that line "*plan*".  Make sense?

The line also is for terraform, so I just excepted the entire bicep directory and all subdirectories.   I did double check, no files other than mine were being affected but it seems like it would save future devs from having to do it.

File: src/bicep/examples/appServicePlan/README.md (hunk 5 above)
Comment (ExchMaster) [reply src/bicep/examples/appServicePlan/README.md:40]: Will come back to this tomorrow night.

File: src/bicep/examples/appServicePlan/modules/appServiceSettings.bicep (hunk 4 above)
Comment (ExchMaster) [reply src/bicep/examples/appServicePlan/modules/appServiceSettings.bicep]: Updated as requested.

File: src/bicep/examples/appServicePlan/modules/appServiceSettings.bicep (hunk 3 above)
Comment (ExchMaster) [reply src/bicep/examples/appServicePlan/modules/appServiceSettings.bicep]: Updated as requested.

File: src/bicep/examples/appServicePlan/modules/appServiceSettings.bicep (hunk 2 above)
Comment (ExchMaster) [reply src/bicep/examples/appServicePlan/modules/appServiceSettings.bicep]: Updated as requested.

File: src/bicep/examples/appServicePlan/appService.bicep (hunk 1 above)
Comment (ExchMaster) [reply src/bicep/examples/appServicePlan/appService.bicep]: Updated in latest commit, good change.   I did opt to leave the 'targetSubscriptionID_Var' in place with a slight modification based on your feedback.   The reason why is because I can pick up the subscription through az-context, so it felt like I was asking the user to do the work twice if that makes sense.   What do you think?

Comment (glennmusa) [reply src/bicep/examples/appServicePlan/appService.bicep]: I think that makes sense 👍

File: src/bicep/examples/appServicePlan/appService.bicep [hunk 8]
```diff
@@ -0,0 +1,50 @@
...
+@description('The name of the web server farm which will be created.  If unchanged or not specified, the MLZ resource prefix + "--ASP" will be utilized.')
+param appServicePlanName string = '${mlzDeploymentVariables.mlzResourcePrefix.Value}-asp'
+
+@description('The name of the resource group in which the app service plan will be deployed. If unchanged or not specified, the MLZ shared services resource group is used.')
+param targetResourceGroup string = '${mlzDeploymentVariables.spokes.Value[2].resourceGroupName}'
+
+@description('If true, enables dynamic scale-in & scale-out based on CPU percentages.  If false, then compute instances remain static with 2 instances supporting all traffic')
+param enableAutoScale bool = true
+
+var targetSubscriptionId_Var = contains(targetResourceGroup, '${mlzDeploymentVariables.spokes.Value[2].resourceGroupName}') ? '${mlzDeploymentVariables.spokes.Value[2].subscriptionId}' : subscription().subscriptionId
```
Comment (glennmusa) [inline src/bicep/examples/appServicePlan/appService.bicep:18]: I think we want an equality comparison here, right? Not just that the parameter contains the same string?

```
var targetSubscriptionId_Var = targetResourceGroup == mlzDeploymentVariables.spokes.Value[2].resourceGroupName ? mlzDeploymentVariables.spokes.Value[2].subscriptionId : subscription().subscriptionId
```

The likelihood is low that it could happen, but an `rg-fizz` is contained in `rg-fizz-buzz`

File: src/bicep/examples/appServicePlan/appService.json [hunk 9]
```diff
@@ -0,0 +1,349 @@
+{
```
Comment (glennmusa) [inline src/bicep/examples/appServicePlan/appService.json:1]: let's update this compiled file to be in sync with `appService.bicep`

File: src/bicep/examples/appServicePlan/README.md (hunk 5 above)
Comment (glennmusa) [reply src/bicep/examples/appServicePlan/README.md:40]: When we've settled on where we want it to live and the contents, I'll push in the BASH Azure CLI equivalents too 👍

```plaintext
az login
cd src/bicep
az deployment sub create -n contoso -f mlz.bicep -l eastus --parameters resourcePrefix=contoso
cd examples
az deployment sub show -n contoso --query properties.outputs >> ./deploymentVariables.json
cd appServicePlan
az deployment sub create -n deployAppServicePlan -f appService.bicep -l eastus
```

File: src/bicep/examples/appServicePlan/appService.bicep (hunk 8 above)
Comment (ExchMaster) [reply src/bicep/examples/appServicePlan/appService.bicep]: Yes sir, completed.

File: src/bicep/examples/appServicePlan/README.md (hunk 5 above)
Comment (ExchMaster) [reply src/bicep/examples/appServicePlan/README.md]: Documentation has been updated as requested.  Thank you Glenn!

File: src/bicep/examples/appServicePlan/README.md [hunk 10]
```diff
@@ -0,0 +1,63 @@
+# App Service Plan Example
+
+This example deploys an App Service Plan (AKA: Web Server Cluster) to support simple web accessible linux docker containers.  It also optionally supports the use of dynamic(up and down) scale settings based on CPU percentage up to a max of 10 compute instances.
```
Comment (brooke-hamilton) [inline src/bicep/examples/appServicePlan/README.md:3]:
```suggestion
This example deploys an App Service Plan (AKA: Web Server Cluster) to support simple web accessible linux docker containers.  It also optionally supports the use of dynamic (up and down) scale settings based on CPU percentage up to a max of 10 compute instances.
```

File: src/bicep/examples/appServicePlan/README.md [hunk 11]
```diff
@@ -0,0 +1,63 @@
...
+
+### Template Parameters
+
+Template Parameters Name | Description
+-----------------------| -----------
+appServicePlanName | The name of the App Service Plan.  If not specified, the name will default to the MLZ default naming pattern.  
+targetResourceGroup | The name of the resource group where the App Service Plan will be deployed.   If not specified, the resource group name will default to the shared services MLZ resource group name and subscription.
+enableAutoScale | A true/false value that determines if dyname auto scale is enabled.  If set to "true", dyanmic auto scale is enabled up to a maximum of 10 compute instances based on CPU percentage exceeding 70% for 10 minutes.   Will also scale in if CPU percentage is below 30% for 10 minutes.  If set to "false", the App Service Plan will statically maintain two compute instances indefinitely.
+
+### Generate MLZ VAriable File (deploymentVariables.json)
```
Comment (brooke-hamilton) [inline src/bicep/examples/appServicePlan/README.md:30]:
```suggestion
### Generate MLZ Variable File (deploymentVariables.json)
```

File: src/bicep/examples/appServicePlan/README.md [hunk 12]
```diff
@@ -0,0 +1,63 @@
...
+
+See below for information on how to create the appropriate deployment variables file for use with this template.
+
+### Template Parameters
+
+Template Parameters Name | Description
+-----------------------| -----------
+appServicePlanName | The name of the App Service Plan.  If not specified, the name will default to the MLZ default naming pattern.  
+targetResourceGroup | The name of the resource group where the App Service Plan will be deployed.   If not specified, the resource group name will default to the shared services MLZ resource group name and subscription.
+enableAutoScale | A true/false value that determines if dyname auto scale is enabled.  If set to "true", dyanmic auto scale is enabled up to a maximum of 10 compute instances based on CPU percentage exceeding 70% for 10 minutes.   Will also scale in if CPU percentage is below 30% for 10 minutes.  If set to "false", the App Service Plan will statically maintain two compute instances indefinitely.
```
Comment (brooke-hamilton) [inline src/bicep/examples/appServicePlan/README.md:28]:
```suggestion
enableAutoScale | A true/false value that determines if dynamic auto scale is enabled.  If set to "true", dynamic auto scale is enabled up to a maximum of 10 compute instances based on CPU percentage exceeding 70% for 10 minutes.   Will also scale down if CPU percentage is below 30% for 10 minutes.  If set to "false", the App Service Plan will statically maintain two compute instances indefinitely.
```

**Review by brooke-hamilton — CHANGES_REQUESTED** [review]
I'm still reviewing but leaving these simple things here while I do deployments.

File: src/bicep/examples/appServicePlan/appService.bicep [hunk 13]
```diff
@@ -0,0 +1,50 @@
...
+param targetResourceGroup string = '${mlzDeploymentVariables.spokes.Value[2].resourceGroupName}'
+
+@description('If true, enables dynamic scale-in & scale-out based on CPU percentages.  If false, then compute instances remain static with 2 instances supporting all traffic')
+param enableAutoScale bool = true
+
+var targetSubscriptionId_Var = targetResourceGroup == '${mlzDeploymentVariables.spokes.Value[2].resourceGroupName}' ? '${mlzDeploymentVariables.spokes.Value[2].subscriptionId}' : subscription().subscriptionId
+var location = deployment().location
+var kind = 'linux'
+var capacity = 2
+var sku = 'premium'
```
Comment (brooke-hamilton) [inline src/bicep/examples/appServicePlan/appService.bicep:22]: The `sku` variable is mapped to the `tier` in the resource provider.  Customers may need to look up valid values for tier in their subscriptions if they get an error because a tier or machine SKU is not available. Maybe remove `tier` altogether as a parameter to the `Microsoft.Web/serverfarms@2020-12-01` provider because it is not required, and repurpose the `sku` parameter to map to the `name` parameter on the `Microsoft.Web/serverfarms@2020-12-01` provider? For example, allow someone to provide `B1` as the `sku` parameter, which would map to the `name` parameter on the provider.

**Review by brooke-hamilton — CHANGES_REQUESTED** [review]

File: src/bicep/examples/deploymentVariables.json [hunk 14]
```diff
@@ -0,0 +1,77 @@
+{
```
[the remark this answers is not in the document]
Comment (ExchMaster) [reply src/bicep/examples/deploymentVariables.json]: Updated gitignore to ignore deploymentVariables.json as requested.

File: src/bicep/examples/appServicePlan/appService.bicep (hunk 13 above)
Comment (ExchMaster) [reply src/bicep/examples/appServicePlan/appService.bicep]: Nice call out, just submitted a commit that updates as requested.  Thank you

**Review by brooke-hamilton — APPROVED** [review]
Looks good. Thanks for the changes. ❤️

**Review by brooke-hamilton — CHANGES_REQUESTED** [review]
Deployed successfully. Minor changes requested.
